go/cmd/ct-fetch/ct-fetch.go

```diff
@@ -278,100 +448,240 @@ func (ld *LogSyncEngine) NewLogWorker(ctLogUrl string) (*LogWorker, error) {
+	// TODO(jms) make the sleep time configurable
+	glog.Infof("[%s] Stopped. Sleeping for 10m", lw.LogURL)
+	select {
+	case <-sigChan:
+		glog.Infof("[%s] Signal caught. Exiting.", lw.LogURL)
+	case <-time.After(10 * time.Second):
+	}
+	signal.Stop(sigChan)
+	close(sigChan)
+}
 
-	if lw.StartPos == lw.EndPos {
-		glog.Infof("[%s] Nothing to do", lw.LogURL)
-		if lw.Bar != nil {
-			lw.Bar.SetTotal((int64)(lw.EndPos), true)
+func (lw *LogWorker) Run(entryChan chan<- CtLogEntry) error {
+	var firstIndex, lastIndex uint64
+
+	switch lw.WorkOrder {
+	case Init:
+		firstIndex = lw.STH.TreeSize - lw.JobSize
+		lastIndex = lw.STH.TreeSize - 1
+		glog.Infof("[%s] Running Init job %d %d", lw.LogURL, firstIndex, lastIndex)
+	case Update:
+		firstIndex = lw.LogState.MaxEntry + 1
+		lastIndex = lw.LogState.MaxEntry + lw.JobSize
+		glog.Infof("[%s] Running Update job %d %d", lw.LogURL, firstIndex, lastIndex)
+	case Backfill:
+		// We will make fewer get-entries requests to the CT Log if we align firstIndex
+		// to a power of two while backfilling.
+		// TODO(jms) document the fact that JobSize should be a power of two
+		if lw.LogState.MinEntry%lw.JobSize != 0 {
+			firstIndex = lw.LogState.MinEntry - (lw.LogState.MinEntry % lw.JobSize)
+		} else {
+			firstIndex = lw.LogState.MinEntry - lw.JobSize
 		}
+		lastIndex = lw.LogState.MinEntry - 1
+		glog.Infof("[%s] Running Backfill job %d %d", lw.LogURL, firstIndex, lastIndex)
+	case Sleep:
+		lw.sleep()
 		return nil
 	}
 
-	finalIndex, finalTime, err := lw.downloadCTRangeToChannel(entryChan)
+	if firstIndex < 0 {
+		firstIndex = 0
+	}
+
+	if lastIndex > lw.STH.TreeSize-1 {
+		lastIndex = lw.STH.TreeSize - 1
+	}
+
+	glog.Infof("[%s] Downloading entries %d through %d",
+		lw.LogURL, firstIndex, lastIndex)
+
+	// We're going to tell users that we downloaded entries
+	// |firstIndex| through |lastIndex|, and we want some assurance
+	// that we've actually done this (especially if we're not getting
+	// entries directly from the CT Log!). We'll ask the log for a
+	// consistency proof between the trees of size |firstIndex| and
+	// |lastIndex+1|. We'll then check that the entries we download
+	// generate the corresponding terms of the proof. We have to handle
+	// the case |firstIndex| = 0 specially.
+	//
+	// Note: we're essentially monitoring the log (as in section 5.3
+	// of RFC 6962). However, we can't always verify the consistency
+	// proofs that we request because we don't always have the
+	// necessary tree heads.
+	//
+	// TODO(jms): Check the proof when |newSize| = |lw.STH.TreeSize|
+	//
+	oldSize := firstIndex
+	newSize := lastIndex + 1
+	if oldSize == 0 {
+		// Special case: the consistency proof with |oldSize| =
+		// 0 is empty. With |oldSize| = 1 (or |oldSize| = 2) it
+		// doesn't include a hash that depends on entry 0 (resp.
+		// 0 or 1). We ensured newSize > 3 when we assigned this
+		// worker its job, so we can use oldSize = 3.
+		oldSize = 3
```

Comment:
Ok - so `oldSize` will never be `0` in `consistencyProofToSubtrees`.

Reply:
Right, thanks for checking.